File: core/src/vct/col/ast/Dereference.java
package vct.col.ast;

public class Dereference extends ASTNode {

  @Override
  public <R,A> R accept_simple(ASTMapping1<R,A> map,A arg){
    return map.map(this,arg);
  }

  
  @Override
  public <T> void accept_simple(ASTVisitor<T> visitor){
    try {
      visitor.visit(this);
    } catch (Throwable t){
      if (thrown.get()!=t){
        System.err.printf("Triggered by %s:%n",getOrigin());
        thrown.set(t);
    }
      throw t;
    }
  }
  
  @Override
  public <T> T accept_simple(ASTMapping<T> map){
    try {
      return map.map(this);
    } catch (Throwable t){
      if (thrown.get()!=t){
        System.err.printf("Triggered by %s:%n",getOrigin());
        thrown.set(t);
     }
      throw t;
    }
  }
 

  public final ASTNode object;
  public final String field;
  
  public Dereference(ASTNode object,String field){
    this.object=object;
    this.field=field;
  }

  public final ASTNode getObject(){
    return object;
  }
  
  public boolean match(ASTNode ast){
    if (ast instanceof Hole){
      return ast.match(this);
    } else if (ast instanceof Dereference){
      Dereference d=(Dereference)ast;
      if (field.equals(d.field)){
        return object.match(d.object);
      } else {
        return false;
      }
    } else {
      return false;
    }
  }
  
}

File: vercors/src/main/java/vct/col/ast/Dereference.scala
package vct.col.ast

import vct.col.util.VisitorHelper

case class Dereference(val obj:ASTNode, val field:String) extends ASTNode with VisitorHelper {
  override def accept_simple[T,A](m:ASTMapping1[T,A], arg:A) = m.map(this, arg)
  override def accept_simple[T](v:ASTVisitor[T]) = handle_standard(() => v.visit(this))
  override def accept_simple[T](m:ASTMapping[T]) = handle_standard(() => m.map(this))
  
  override def `match`(ast:ASTNode) = ast match {
    case Dereference(o, `field`) => o `match` obj
    case h:Hole => h `match` this
    case _ => false
  }
}
